core/executor/src/wasmtime/runtime.rs

```diff
@@ -0,0 +1,385 @@
+		(memory_plan.memory.minimum, memory_plan.memory.maximum)
+	};
+
+	// Check that heap_pages is within the allowed range.
+	let max_heap_pages = max_memory_size.map(|max| max.saturating_sub(min_memory_size));
+	let heap_pages = heap_pages_valid(heap_pages, max_heap_pages)
+		.ok_or_else(|| WasmError::InvalidHeapPages)?;
+
+	// Call to determine runtime version.
+	let version_result = call_method(
+		&mut context,
+		&mut compiled_module,
+		ext,
+		"Core_version",
+		&[],
+		heap_pages
+	);
+	let version = version_result
+		.ok()
+		.and_then(|v| RuntimeVersion::decode(&mut v.as_slice()).ok());
+	Ok(WasmtimeRuntime {
+		module: compiled_module,
+		context,
+		max_heap_pages,
+		heap_pages,
+		version,
+	})
+}
+
+fn create_compiled_unit(code: &[u8])
+	-> std::result::Result<(CompiledModule, Context), WasmError>
+{
+	let isa = target_isa()?;
+	let mut context = Context::with_isa(isa, CompilationStrategy::Cranelift);
+
+	// Enable/disable producing of debug info.
+	context.set_debug_info(false);
+
+	// Instantiate and link the env module.
+	let global_exports = context.get_global_exports();
+	let env_module = instantiate_env_module(global_exports)?;
+	context.name_instance("env".to_owned(), env_module);
+
+	// Compile the wasm module.
+	let module = context.compile_module(&code)
+		.map_err(WasmError::WasmtimeSetup)?;
+
+	Ok((module, context))
+}
+
+/// Call a function inside a precompiled Wasm module.
+fn call_method(
+	context: &mut Context,
+	module: &mut CompiledModule,
+	ext: &mut dyn Externalities,
+	method: &str,
+	data: &[u8],
+	heap_pages: u32,
+) -> Result<Vec<u8>> {
+	// Old exports get clobbered in `InstanceHandle::new` if we don't explicitly remove them first.
+	//
+	// The global exports mechanism is temporary in Wasmtime and expected to be removed.
+	// https://github.com/CraneStation/wasmtime/issues/332
+	clear_globals(&mut *context.get_global_exports().borrow_mut());
+
+	let mut instance = module.instantiate()
+		.map_err(SetupError::Instantiate)
+		.map_err(ActionError::Setup)
+		.map_err(Error::Wasmtime)?;
+
+	// Ideally there would be a way to set the heap pages during instantiation rather than
+	// growing the memory after the fact. Current this may require an additional mmap and copy.
+	// However, the wasmtime API doesn't support modifying the size of memory on instantiation
+	// at this time.
+	grow_memory(&mut instance, heap_pages)?;
+
+	// Initialize the function executor state.
+	let executor_state = FunctionExecutorState::new(
+		// Unsafely extend the reference lifetime to static. This is necessary because the
+		// host state must be `Any`. This is OK as we will drop this object either before
+		// exiting the function in the happy case or in the worst case on the next runtime
+		// call, which also resets the host state. Thus this reference can never actually
+		// outlive the Context.
+		unsafe { mem::transmute::<_, &'static mut Compiler>(context.compiler_mut()) },
```

Comment:
Could you please write down the first type explictly? I know that it might seem a bit verbose, but omitting one of the types in `transmute` is considered to be a source of errors. Also, IMO, it would slightly improve readibility.